docs/package_templates/cmake_package/all/conanfile.py

```diff
@@ -86,46 +86,51 @@ def validate(self):
+        self.tool_requires("cmake/[>=3.16 <4]")
 
     def source(self):
         get(self, **self.conan_data["sources"][self.version], strip_root=True)
 
     def generate(self):
-        # BUILD_SHARED_LIBS and POSITION_INDEPENDENT_CODE are automatically parsed when self.options.shared or self.options.fPIC exist
+        # BUILD_SHARED_LIBS and POSITION_INDEPENDENT_CODE are set automatically as tc.variables when self.options.shared or self.options.fPIC exist
+        # Note that tc.variables require either cmake_minimum_required() >= 3.13 or the CMP0077 policy set to NEW to work correctly.
         tc = CMakeToolchain(self)
         # Boolean values are preferred instead of "ON"/"OFF"
-        tc.variables["PACKAGE_CUSTOM_DEFINITION"] = True
+        tc.variables["PACKAGE_BUILD_TESTS"] = False
         if is_msvc(self):
-            # don't use self.settings.compiler.runtime
             tc.variables["USE_MSVC_RUNTIME_LIBRARY_DLL"] = not is_msvc_static_runtime(self)
-        # deps_cpp_info, deps_env_info and deps_user_info are no longer used
         if self.dependencies["dependency"].options.foobar:
-            tc.variables["DEPENDENCY_LIBPATH"] = self.dependencies["dependency"].cpp_info.libdirs
-        # cache_variables should be used sparingly, example setting cmake policies
+            tc.variables["DEPENDENCY_LIBPATH"] = self.dependencies["dependency"].cpp_info.libdir.replace("\\", "/")
+        # cache_variables should be used sparingly, for example to set CMake policies
         tc.cache_variables["CMAKE_POLICY_DEFAULT_CMP0077"] = "NEW"
         tc.generate()
-        # In case there are dependencies listed on requirements, CMakeDeps should be used
-        tc = CMakeDeps(self)
-        tc.generate()
-        # In case there are dependencies listed on build_requirements, VirtualBuildEnv should be used
-        tc = VirtualBuildEnv(self)
-        tc.generate(scope="build")
+
+        # In case there are dependencies listed under requirements, CMakeDeps should be used
+        deps = CMakeDeps(self)
+        # You can override the CMake package and target names if they don't match the names used in the project
+        deps.set_property("fontconfig", "cmake_file_name", "Fontconfig")
+        deps.set_property("fontconfig", "cmake_target_name", "Fontconfig::Fontconfig")
+        deps.generate()
+
+        # In case there are dependencies listed under build_requirements, VirtualBuildEnv should be used
+        VirtualBuildEnv(self).generate()
 
     def _patch_sources(self):
         apply_conandata_patches(self)
         # remove bundled xxhash
-        rm(self, "whateer.*", os.path.join(self.source_folder, "lib"))
+        rm(self, "whatever.*", os.path.join(self.source_folder, "lib"))
         replace_in_file(self, os.path.join(self.source_folder, "CMakeLists.txt"), "...", "")
+        # disable subdirectories by truncating their CMakeLists.txt
+        save(self, os.path.join(self.source_folder, "tests", "CMakeLists.txt"), "")
```

Comment:
`test` and `examples` directories need to be disabled quite often and this approach requires less maintenance overhead compared to patching `CMakeLists.txt`.

Reply:
I would not include it. It encourages people doing more patches instead of trying to check it with the upstream an alternative. Remember, patches are the last resource always.

Reply:
My worry with this as explained elsewhere is that in the case where a new version changes how their tests etc are set up, we won't fail as it would have happened with patches, thus making it easier to sneak tests and the like without noticing. wdyt?

Reply:
Why not both - use this as a patch and submit a fix upstream as well? :grin: 

Using plain patches is better in principle, but the `add_subdirectory()` statements are typically one of the very last ones in a `CMakeLists.txt` file, which for larger projects means that this needs to be updated for almost every single release. I can remove this example here, but in general, would a `replace_in_file()` with `strict=True` be acceptable instead?